EtcdConfigurationSource

spoon16 opened the pull request:
An Etcd watched configuration source very similar to the ZooKeeperConfigurationSource.

Uses the Boon Etcd client https://github.com/boonproject/boon/blob/master/etcd/README.md.  Seems to be the most widely adopted Etcd client.

Tests included.

cloudbees-pull-request-builder commented:
[NetflixOSS » archaius » archaius-pull-requests #85](https://netflixoss.ci.cloudbees.com/job/NetflixOSS/job/archaius/job/archaius-pull-requests/85/) FAILURE
Looks like there's a problem with this pull request

spoon16 commented:
Build failed because of a JVM error... or maybe a bunch of failing unit tests unrelated to the Etcd module.

Unless I am missing something.

cloudbees-pull-request-builder commented:
[archaius-pull-requests #254](https://netflixoss.ci.cloudbees.com/job/archaius-pull-requests/254/) SUCCESS
This pull request looks good

howardyuan commented:
There're two pull-requests builds. One of them was not configured with
proper Gradle opts so it runs out Permgen. I'll see how to either disable
the second one or add proper GRADLE_OPTS.

On Thu, Jun 11, 2015 at 6:02 PM, Eric Schoonover <notifications@github.com>
wrote:

> Build failed because of a JVM error... or maybe a bunch of failing unit
> tests unrelated to the Etcd module.
>
> Unless I am missing something.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Netflix/archaius/pull/318#issuecomment-111318271>.
>

howardyuan commented on archaius-etcd/src/main/java/com/netflix/config/source/EtcdConfigurationSource.java:
```diff
@@ -0,0 +1,147 @@
+            final Map<String, Object> set = Maps.newHashMap();
+            final Map<String, Object> delete = Maps.newHashMap();
+
+            final String action = updateResponse.action().toLowerCase();
```
Can the action() return null when it's error?

spoon16 replied:
only if I duplicate the etcd.waitRecursive call on l:89

It'd probably be best to capture the wasError condition, throw an exception and have a finally that calls etcd.waitRecursive.  But I'm not sure that really makes things more readable.  I'll just duplicate l:89 and return early.

howardyuan commented on archaius-etcd/src/main/java/com/netflix/config/source/EtcdConfigurationSource.java:
```diff
@@ -0,0 +1,147 @@
+                logger.error("Etcd failed with an error response: %s", updateResponse);
+            }
+
+            final Map<String, Object> create = Maps.newHashMap();
```
I was a little puzzled by three maps here for a while. For sure, we will get two empty maps and one map with one entry in there? Not really sure if I have a better idea yet, just want to point out.

spoon16 replied:
Yeah, not sure what would be better here. I could do something like:

```java

```

howardyuan commented on archaius-etcd/src/main/java/com/netflix/config/source/EtcdConfigurationSource.java:
```diff
@@ -0,0 +1,147 @@
+                final String etcdKey = node.key();
+                final String sourceKey = Iterables.getLast(keySplitter.split(etcdKey));
+                final String value = node.getValue();
+                valueCache.put(sourceKey, value);
```
what if it's a delete, then should you remove the entry from the cache?

spoon16 replied:
good catch, will fix

spoon16 replied:
fixed and added a test to validate

cloudbees-pull-request-builder commented:
[NetflixOSS » archaius » archaius-pull-requests #86](https://netflixoss.ci.cloudbees.com/job/NetflixOSS/job/archaius/job/archaius-pull-requests/86/) FAILURE
Looks like there's a problem with this pull request

cloudbees-pull-request-builder commented:
[archaius-pull-requests #255](https://netflixoss.ci.cloudbees.com/job/archaius-pull-requests/255/) SUCCESS
This pull request looks good

spoon16 commented:
@howardyuan, tried to address your feedback with the following changes.

- fixed bug where delete wouldn't remove the value from the `EtcdConfigurationSource`
- added a unit test to validate delete behavior
- refactored the `updateHandler` in `EtcdConfigurationSource`... tried to improve readability by not always instantiating three maps

howardyuan commented:
Looks good to me. Will leave the PR open for a few days to see if anybody else have any more comments.

chandrapersonal commented:
can i pull the changes.

spoon16 commented:
@howardyuan merge?

chandrapersonal commented:
why there is no case for node action "compareandswap" in etcd UpdateHandler. we have node actions create,set and delete action on UpdateHandler . but no  compareandswap. Is there any reason for not have that action.

spoon16 commented:
@chandrapersonal I don't think "compareandswap" is a type of update.  It's part of the Etcd update API and the side effect is that an update notification is emitted, either set or create.
